diff -Naur -x 'mpfr/*' -x 'gmp/*' -x '*.xz' -x .gitignore -x '.settings/*' -x .project -x .cproject -x '*~' -x '*.o' -x '*.dll' -x '*.exe' gcc.orig/.settings/language.settings.xml gcc/.settings/language.settings.xml
--- gcc.orig/boehm-gc/configure	2017-04-12 19:33:03.000000000 +0200
+++ gcc/boehm-gc/configure	2017-04-15 18:04:52.722500000 +0200
@@ -11053,7 +11053,7 @@
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-ldl  $LIBS"
+LIBS="-ldl -lpsapi $LIBS"
 if test x$gcc_no_link = xyes; then
   as_fn_error "Link tests are not allowed after GCC_NO_EXECUTABLES." "$LINENO" 5
 fi
@@ -11087,7 +11087,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_dl_dlopen" >&5
 $as_echo "$ac_cv_lib_dl_dlopen" >&6; }
 if test "x$ac_cv_lib_dl_dlopen" = x""yes; then :
-  lt_cv_dlopen="dlopen" lt_cv_dlopen_libs="-ldl"
+  lt_cv_dlopen="dlopen" lt_cv_dlopen_libs="-ldl -lpsapi"
 else
 
     lt_cv_dlopen="dyld"
@@ -11155,7 +11155,7 @@
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-ldl  $LIBS"
+LIBS="-ldl -lpsapi $LIBS"
 if test x$gcc_no_link = xyes; then
   as_fn_error "Link tests are not allowed after GCC_NO_EXECUTABLES." "$LINENO" 5
 fi
@@ -15013,7 +15013,7 @@
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-ldl  $LIBS"
+LIBS="-ldl -lpsapi $LIBS"
 if test x$gcc_no_link = xyes; then
   as_fn_error "Link tests are not allowed after GCC_NO_EXECUTABLES." "$LINENO" 5
 fi
@@ -15047,7 +15047,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_dl_dlopen" >&5
 $as_echo "$ac_cv_lib_dl_dlopen" >&6; }
 if test "x$ac_cv_lib_dl_dlopen" = x""yes; then :
-  EXTRA_TEST_LIBS="$EXTRA_TEST_LIBS -ldl"
+  EXTRA_TEST_LIBS="$EXTRA_TEST_LIBS -ldl -lpsapi"
 fi
 
     ;;
diff -Naur -x 'mpfr/*' -x 'gmp/*' -x '*.xz' -x .gitignore -x '.settings/*' -x .project -x .cproject -x '*~' -x '*.o' -x '*.dll' -x '*.exe' gcc.orig/gcc/c/Make-lang.in gcc/gcc/c/Make-lang.in
--- gcc.orig/gcc/c/Make-lang.in	2017-04-12 19:33:04.000000000 +0200
+++ gcc/gcc/c/Make-lang.in	2017-04-16 20:10:20.171300100 +0200
@@ -69,7 +69,7 @@
 
 cc1$(exeext): $(C_OBJS) cc1-checksum.o $(BACKEND) $(LIBDEPS)
 	+$(LLINKER) $(ALL_LINKERFLAGS) $(LDFLAGS) -o $@ $(C_OBJS) \
-	  cc1-checksum.o $(BACKEND) $(LIBS) $(BACKENDLIBS)
+	  cc1-checksum.o $(BACKEND) $(LIBS) $(BACKENDLIBS) -Wl,--output-def,cc1.def,--out-implib,libcc1.a
 #
 # Build hooks:
 
diff -Naur -x 'mpfr/*' -x 'gmp/*' -x '*.xz' -x .gitignore -x '.settings/*' -x .project -x .cproject -x '*~' -x '*.o' -x '*.dll' -x '*.exe' gcc.orig/gcc/configure gcc/gcc/configure
--- gcc.orig/gcc/configure	2017-04-12 19:33:04.000000000 +0200
+++ gcc/gcc/configure	2017-04-15 18:39:20.091000000 +0200
@@ -18206,7 +18206,7 @@
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-ldl  $LIBS"
+LIBS="-ldl -lpsapi $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -18237,7 +18237,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_dl_dlopen" >&5
 $as_echo "$ac_cv_lib_dl_dlopen" >&6; }
 if test "x$ac_cv_lib_dl_dlopen" = x""yes; then :
-  lt_cv_dlopen="dlopen" lt_cv_dlopen_libs="-ldl"
+  lt_cv_dlopen="dlopen" lt_cv_dlopen_libs="-ldl -lpsapi"
 else
 
     lt_cv_dlopen="dyld"
@@ -18302,7 +18302,7 @@
   $as_echo_n "(cached) " >&6
 else
   ac_check_lib_save_LIBS=$LIBS
-LIBS="-ldl  $LIBS"
+LIBS="-ldl -lpsapi $LIBS"
 cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
 
@@ -18333,7 +18333,7 @@
 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $ac_cv_lib_dl_dlopen" >&5
 $as_echo "$ac_cv_lib_dl_dlopen" >&6; }
 if test "x$ac_cv_lib_dl_dlopen" = x""yes; then :
-  lt_cv_dlopen="dlopen" lt_cv_dlopen_libs="-ldl"
+  lt_cv_dlopen="dlopen" lt_cv_dlopen_libs="-ldl -lpsapi"
 else
   { $as_echo "$as_me:${as_lineno-$LINENO}: checking for dlopen in -lsvld" >&5
 $as_echo_n "checking for dlopen in -lsvld... " >&6; }
@@ -29245,10 +29245,14 @@
 	 ${CC} ${CFLAGS} ${LDFLAGS} -rdynamic conftest.c -o conftest$ac_exeext > /dev/null 2>&1
 	 if $export_sym_check conftest$ac_exeext | grep -q foobar > /dev/null; then
 	   plugin_rdynamic=yes
-	   pluginlibs="-rdynamic"
+           enable_plugin=yes
+	   pluginlibs="-Wl,--export-all-symbols"
+           $as_echo "#define HAVE_DLFCN_H 1" >>confdefs.h
 	 else
-	   plugin_rdynamic=no
-	   enable_plugin=no
+           plugin_rdynamic=yes
+           enable_plugin=yes
+           pluginlibs="-Wl,--export-all-symbols"
+           $as_echo "#define HAVE_DLFCN_H 1" >>confdefs.h
 	 fi
 	 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $plugin_rdynamic" >&5
 $as_echo "$plugin_rdynamic" >&6; }
@@ -29317,7 +29321,7 @@
 fi
 
      if test x"$ac_cv_search_dlopen" = x"-ldl"; then
-       pluginlibs="$pluginlibs -ldl"
+       pluginlibs="$pluginlibs -ldl -lpsapi"
      fi
      LIBS="$saved_LIBS"
 
@@ -29339,7 +29343,7 @@
 $as_echo_n "checking for -fPIC -shared... " >&6; }
      cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
-extern int X;
+int X;
 int
 main ()
 {
diff -Naur -x 'mpfr/*' -x 'gmp/*' -x '*.xz' -x .gitignore -x '.settings/*' -x .project -x .cproject -x '*~' -x '*.o' -x '*.dll' -x '*.exe' gcc.orig/gcc/cp/Make-lang.in gcc/gcc/cp/Make-lang.in
--- gcc.orig/gcc/cp/Make-lang.in	2017-04-12 19:33:04.000000000 +0200
+++ gcc/gcc/cp/Make-lang.in	2017-04-16 20:12:32.045300100 +0200
@@ -98,7 +98,7 @@
 
 cc1plus$(exeext): $(CXX_OBJS) cc1plus-checksum.o $(BACKEND) $(LIBDEPS)
 	+$(LLINKER) $(ALL_LINKERFLAGS) $(LDFLAGS) -o $@ \
-	      $(CXX_OBJS) cc1plus-checksum.o $(BACKEND) $(LIBS) $(BACKENDLIBS)
+	      $(CXX_OBJS) cc1plus-checksum.o $(BACKEND) $(LIBS) $(BACKENDLIBS) -Wl,--output-def,cc1plus.def,--out-implib,libcc1plus.a
 
 ifeq ($(ENABLE_MAINTAINER_RULES), true)
 # Special build rule.  This is a maintainer rule, that is only
@@ -187,6 +187,7 @@
 	      ( cd $(DESTDIR)$(bindir) && \
 		$(LN) $(CXX_INSTALL_NAME)$(exeext) $(CXX_TARGET_INSTALL_NAME)$(exeext) ); \
 	    fi ; \
+	    $(INSTALL) cc1plus.def libcc1plus.a $(DESTDIR)$(libdir); \
 	  fi; \
 	fi
 
diff -Naur -x 'mpfr/*' -x 'gmp/*' -x '*.xz' -x .gitignore -x '.settings/*' -x .project -x .cproject -x '*~' -x '*.o' -x '*.dll' -x '*.exe' gcc.orig/gcc/fortran/Make-lang.in gcc/gcc/fortran/Make-lang.in
--- gcc.orig/gcc/fortran/Make-lang.in	2017-04-12 19:33:04.000000000 +0200
+++ gcc/gcc/fortran/Make-lang.in	2017-04-14 16:17:39.522000000 +0200
@@ -96,7 +96,7 @@
 		$(BACKEND) $(LIBDEPS) attribs.o
 	+$(LLINKER) $(ALL_LINKERFLAGS) $(LDFLAGS) -o $@ \
 		$(F95_OBJS) $(BACKEND) $(ZLIB) $(LIBS) attribs.o \
-		$(BACKENDLIBS)
+		$(BACKENDLIBS) -Wl,--output-def,f951.def,--out-implib,libf951.a
 
 gt-fortran-trans.h    : s-gtype; @true
 #
diff -Naur -x 'mpfr/*' -x 'gmp/*' -x '*.xz' -x .gitignore -x '.settings/*' -x .project -x .cproject -x '*~' -x '*.o' -x '*.dll' -x '*.exe' gcc.orig/gcc/go/Make-lang.in gcc/gcc/go/Make-lang.in
--- gcc.orig/gcc/go/Make-lang.in	2017-04-12 19:33:04.000000000 +0200
+++ gcc/gcc/go/Make-lang.in	2017-04-14 16:32:57.471000000 +0200
@@ -36,7 +36,7 @@
 gccgo$(exeext): $(GCCGO_OBJS) $(EXTRA_GCC_OBJS) libcommon-target.a $(LIBDEPS)
 	+$(LINKER) $(ALL_LINKERFLAGS) $(LDFLAGS) -o $@ \
 	  $(GCCGO_OBJS) $(EXTRA_GCC_OBJS) libcommon-target.a \
-	  $(EXTRA_GCC_LIBS) $(LIBS)
+	  $(EXTRA_GCC_LIBS) $(LIBS) -Wl,--output-def,gccgo.def,--out-implib,libgccgo.a
 
 # The cross-compiler version.  This is built mainly as a signal to the
 # go.install-common target.  If this executable exists, it means that
diff -Naur -x 'mpfr/*' -x 'gmp/*' -x '*.xz' -x .gitignore -x '.settings/*' -x .project -x .cproject -x '*~' -x '*.o' -x '*.dll' -x '*.exe' gcc.orig/gcc/lto/Make-lang.in gcc/gcc/lto/Make-lang.in
--- gcc.orig/gcc/lto/Make-lang.in	2017-04-12 19:33:04.000000000 +0200
+++ gcc/gcc/lto/Make-lang.in	2017-04-14 16:31:24.690000000 +0200
@@ -69,7 +69,7 @@
 
 $(LTO_EXE): $(LTO_OBJS) $(BACKEND) $(LIBDEPS)
 	+$(LLINKER) $(ALL_LINKERFLAGS) $(LDFLAGS) -o $@ \
-		$(LTO_OBJS) $(BACKEND) $(BACKENDLIBS) $(LIBS)
+		$(LTO_OBJS) $(BACKEND) $(BACKENDLIBS) $(LIBS) -Wl,--output-def,lto.def,--out-implib,liblto.a
 
 # LTO testing is done as part of C/C++/Fortran etc. testing.
 check-lto:
diff -Naur -x 'mpfr/*' -x 'gmp/*' -x '*.xz' -x .gitignore -x '.settings/*' -x .project -x .cproject -x '*~' -x '*.o' -x '*.dll' -x '*.exe' gcc.orig/gcc/objc/Make-lang.in gcc/gcc/objc/Make-lang.in
--- gcc.orig/gcc/objc/Make-lang.in	2017-04-12 19:33:05.000000000 +0200
+++ gcc/gcc/objc/Make-lang.in	2017-04-14 16:30:41.956000000 +0200
@@ -65,7 +65,7 @@
 cc1obj$(exeext): $(OBJC_OBJS) $(C_AND_OBJC_OBJS) cc1obj-checksum.o $(BACKEND) $(LIBDEPS)
 	+$(LLINKER) $(ALL_LINKERFLAGS) $(LDFLAGS) -o $@ \
 	      $(OBJC_OBJS) $(C_AND_OBJC_OBJS) cc1obj-checksum.o \
-	      $(BACKEND) $(LIBS) $(BACKENDLIBS)
+	      $(BACKEND) $(LIBS) $(BACKENDLIBS) -Wl,--output-def,cc1obj.def,--out-implib,libcc1obj.a
 
 objc.srcextra:
 
diff -Naur -x 'mpfr/*' -x 'gmp/*' -x '*.xz' -x .gitignore -x '.settings/*' -x .project -x .cproject -x '*~' -x '*.o' -x '*.dll' -x '*.exe' gcc.orig/gcc/objcp/Make-lang.in gcc/gcc/objcp/Make-lang.in
--- gcc.orig/gcc/objcp/Make-lang.in	2017-04-12 19:33:05.000000000 +0200
+++ gcc/gcc/objcp/Make-lang.in	2017-04-14 16:29:57.450000000 +0200
@@ -68,7 +68,8 @@
 
 cc1objplus$(exeext): $(OBJCXX_OBJS) cc1objplus-checksum.o $(BACKEND) $(LIBDEPS)
 	+$(LLINKER) $(ALL_LINKERFLAGS) $(LDFLAGS) -o $@ \
-		$(OBJCXX_OBJS) cc1objplus-checksum.o $(BACKEND) $(LIBS) $(BACKENDLIBS)
+		$(OBJCXX_OBJS) cc1objplus-checksum.o $(BACKEND) $(LIBS) \
+		$(BACKENDLIBS) -Wl,--output-def,cc1objplus.def,--out-implib,libcc1objplus.a
 
 # Objective C++ language specific files.
 
diff -Naur -x 'mpfr/*' -x 'gmp/*' -x '*.xz' -x .gitignore -x '.settings/*' -x .project -x .cproject -x '*~' -x '*.o' -x '*.dll' -x '*.exe' gcc.orig/gcc/plugin.c gcc/gcc/plugin.c
--- gcc.orig/gcc/plugin.c	2017-04-12 19:33:05.000000000 +0200
+++ gcc/gcc/plugin.c	2017-04-16 20:01:44.078300100 +0200
@@ -570,6 +570,8 @@
   const char *err;
   PTR_UNION_TYPE (plugin_init_func) plugin_init_union;
 
+  printf("try_init_one_plugin\n");
+
   /* We use RTLD_NOW to accelerate binding and detect any mismatch
      between the API expected by the plugin and the GCC API; we use
      RTLD_GLOBAL which is useful to plugins which themselves call
diff -Naur -x 'mpfr/*' -x 'gmp/*' -x '*.xz' -x .gitignore -x '.settings/*' -x .project -x .cproject -x '*~' -x '*.o' -x '*.dll' -x '*.exe' gcc.orig/gcc/system.h gcc/gcc/system.h
--- gcc.orig/gcc/system.h	2017-04-12 19:33:05.000000000 +0200
+++ gcc/gcc/system.h	2017-04-15 18:25:19.953500000 +0200
@@ -673,7 +673,7 @@
 
 #if defined (ENABLE_PLUGIN) && defined (HAVE_DLFCN_H)
 /* If plugin support is enabled, we could use libdl.  */
-#include <dlfcn.h>
+#include "dlfcn.h" // FIXME: use better way of including the dlfcn-win32 header
 #endif
 
 /* Do not introduce a gmp.h dependency on the build system.  */
diff -Naur -x 'mpfr/*' -x 'gmp/*' -x '*.xz' -x .gitignore -x '.settings/*' -x .project -x .cproject -x '*~' -x '*.o' -x '*.dll' -x '*.exe' gcc.orig/gmp/.gdbinit gcc/gmp/.gdbinit
--- gcc.orig/libcc1/configure	2017-04-12 19:33:09.000000000 +0200
+++ gcc/libcc1/configure	2017-04-15 19:20:07.422000000 +0200
@@ -14571,10 +14571,13 @@
 	 ${CC} ${CFLAGS} ${LDFLAGS} -rdynamic conftest.c -o conftest$ac_exeext > /dev/null 2>&1
 	 if $export_sym_check conftest$ac_exeext | grep -q foobar > /dev/null; then
 	   plugin_rdynamic=yes
-	   pluginlibs="-rdynamic"
+           pluginlibs="-Wl,--export-all-symbols"
+           $as_echo "#define HAVE_DLFCN_H 1" >>confdefs.h
 	 else
-	   plugin_rdynamic=no
-	   enable_plugin=no
+	   plugin_rdynamic=yes
+           pluginlibs="-Wl,--export-all-symbols"
+	   enable_plugin=yes
+           $as_echo "#define HAVE_DLFCN_H 1" >>confdefs.h
 	 fi
 	 { $as_echo "$as_me:${as_lineno-$LINENO}: result: $plugin_rdynamic" >&5
 $as_echo "$plugin_rdynamic" >&6; }
@@ -14585,7 +14588,7 @@
      fi
 
      # Check -ldl
-     saved_LIBS="$LIBS"
+     saved_LIBS="$LIBS -ldl -lpsapi"
      { $as_echo "$as_me:${as_lineno-$LINENO}: checking for library containing dlopen" >&5
 $as_echo_n "checking for library containing dlopen... " >&6; }
 if test "${ac_cv_search_dlopen+set}" = set; then :
@@ -14610,7 +14613,7 @@
   return 0;
 }
 _ACEOF
-for ac_lib in '' dl; do
+for ac_lib in '' dl psapi; do
   if test -z "$ac_lib"; then
     ac_res="none required"
   else
@@ -14643,7 +14646,7 @@
 fi
 
      if test x"$ac_cv_search_dlopen" = x"-ldl"; then
-       pluginlibs="$pluginlibs -ldl"
+       pluginlibs="$pluginlibs -ldl -lpsapi"
      fi
      LIBS="$saved_LIBS"
 
@@ -14665,7 +14668,7 @@
 $as_echo_n "checking for -fPIC -shared... " >&6; }
      cat confdefs.h - <<_ACEOF >conftest.$ac_ext
 /* end confdefs.h.  */
-extern int X;
+int X;
 int
 main ()
 {
diff -Naur -x 'mpfr/*' -x 'gmp/*' -x '*.xz' -x .gitignore -x '.settings/*' -x .project -x .cproject -x '*~' -x '*.o' -x '*.dll' -x '*.exe' gcc.orig/libjava/classpath/.cvsignore gcc/libjava/classpath/.cvsignore
